### PR TITLE
Fix processing of proxybypass variables from environment

### PR DESCRIPTION
**Pull request by EzzhevNikita**
This PR should address the issue with the processing of proxy bypass variables. Current logic incorrectly processes any host that is passed to NO_PROXY env variable in the format: _.somehost.net_. Such hosts will be wrongly interpreted because they contain points in hosts. Point is a special symbol for regex, so it should be escaped by backslash.

### CODE REVIEW

File: lib/Util.ts [hunk 1]
```diff
@@ -107,17 +107,17 @@ export async function decompressGzippedContent(buffer: Buffer, charset?: string)
...
-        // We need to keep this around for back-compat purposes
-        return new RegExp(bypass, 'i')    
-    }
-    catch(err) {
-        if (err instanceof SyntaxError && (bypass || "").startsWith("*")) {            
-            let wildcardEscaped = bypass.replace('*', '(.*)');
-            return new RegExp(wildcardEscaped, 'i');
-        }
-        throw err;
+    let searchRegExpToReplaceSpecialChars = new RegExp('([.])', 'g');
```
Comment (alexander-smolyakov) [inline lib/Util.ts:110]:
``` suggestion
    const searchRegExpToReplaceSpecialChars: RegExp = new RegExp('([.])', 'g');
```

File: lib/Util.ts [hunk 2]
```diff
@@ -107,17 +107,17 @@ export async function decompressGzippedContent(buffer: Buffer, charset?: string)
...
+
+    // replace all . symbols in string by \. because point is a special character
+    let safeRegex = (bypass || "").replace(searchRegExpToReplaceSpecialChars, '\\$1');
```
Comment (alexander-smolyakov) [inline lib/Util.ts:113]:
```suggestion
    let safeRegex: string = (bypass || "").replace(searchRegExpToReplaceSpecialChars, '\\$1');
```

File: lib/Util.ts [hunk 3]
```diff
@@ -107,17 +107,17 @@ export async function decompressGzippedContent(buffer: Buffer, charset?: string)
...
+    
+    // check if expression starts with asterisk and replace it with .*
+    if (safeRegex && safeRegex.startsWith("*")) {
+        safeRegex = safeRegex.replace("*", ".*");
     }
+
+    return new RegExp(safeRegex, 'i');
```
Comment (alexander-smolyakov) [inline lib/Util.ts:120]: Could it make sense to expand our test coverage for this function?

https://github.com/microsoft/typed-rest-client/blob/9397d2e164afe5f70119dde75935dba9d09d7afd/test/units/utiltests.ts#L17-L44

Comment (EzzhevNikita) [reply lib/Util.ts:120]: Yeah, make sense

Comment (EzzhevNikita) [reply lib/Util.ts:120]: Added several tests that cover cases that I am trying to fix by this PR

**Review by alexander-smolyakov — APPROVED** [review]
LGTM, thanks!

File: lib/Util.ts [hunk 4]
```diff
@@ -107,17 +107,17 @@ export async function decompressGzippedContent(buffer: Buffer, charset?: string)
...
-    catch(err) {
-        if (err instanceof SyntaxError && (bypass || "").startsWith("*")) {            
-            let wildcardEscaped = bypass.replace('*', '(.*)');
-            return new RegExp(wildcardEscaped, 'i');
-        }
-        throw err;
+    const searchRegExpToReplaceSpecialChars: RegExp = new RegExp('([.])', 'g');
+
+    // replace all . symbols in string by \. because point is a special character
+    let safeRegex: string = (bypass || "").replace(searchRegExpToReplaceSpecialChars, '\\$1');
```
Comment (anatolybolshakov) [inline lib/Util.ts:113]: Won't it break customers which use regexp as a value with dots as special characters?

Comment (EzzhevNikita) [reply lib/Util.ts:113]: Generally NO_PROXY env variable shouldn't contain a regex inside, the strings the set inside should be interpreted as is. The reason why we need to escape dot symbols is that we look for the specific domains in URLs using the RegExp test function which accepts a regular expression as a pattern to search, so we transform the string to the RegExp. NO_PROXY is quite a common way of setup a proxy bypass list and as far as I see there is no soft that supports regexp inside, so dots shouldn't be used as special characters there and in my mind, they could be safely escaped.

Comment (EzzhevNikita) [reply lib/Util.ts:113]: And also our docs related to the setup of proxy bypass domains is saying that dots should be escaped - https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/proxy?view=azure-devops&tabs=windows#specify-proxy-bypass-urls

File: lib/Util.ts [hunk 5]
```diff
@@ -107,17 +107,17 @@ export async function decompressGzippedContent(buffer: Buffer, charset?: string)
...
-        throw err;
+    const searchRegExpToReplaceSpecialChars: RegExp = new RegExp('(?<!\\\\)([.])(?!\\*)', 'g');
+
+    // check if expression starts with asterisk and replace it with .*
+    if (bypass && bypass.startsWith("*")) {
+        bypass = bypass.replace("*", ".*");
     }
+
+    // replace all . symbols in string by \. because point is a special character
+    var safeRegex = (bypass || "").replace(searchRegExpToReplaceSpecialChars, '\\$1');
```
Comment (alexander-smolyakov) [inline lib/Util.ts:118]: Nit: Use `let` or `const` please

**Review by alexander-smolyakov — APPROVED** [review]
LGTM and please take a look at the comment

File: lib/Util.ts (hunk 5 above)
Comment (EzzhevNikita) [reply lib/Util.ts]: done

File: lib/Util.ts [hunk 6]
```diff
@@ -107,17 +107,17 @@ export async function decompressGzippedContent(buffer: Buffer, charset?: string)
...
-        // We need to keep this around for back-compat purposes
-        return new RegExp(bypass, 'i')    
-    }
-    catch(err) {
-        if (err instanceof SyntaxError && (bypass || "").startsWith("*")) {            
-            let wildcardEscaped = bypass.replace('*', '(.*)');
-            return new RegExp(wildcardEscaped, 'i');
-        }
-        throw err;
+    const searchRegExpToReplaceSpecialChars: RegExp = new RegExp('(?<!\\\\)([.])(?!\\*)', 'g');
```
Comment (anatolybolshakov) [inline lib/Util.ts:110]: Nit: could it make sense to move it to constants and not initialize on every call?

**Review by anatolybolshakov — APPROVED** [review]

File: lib/Util.ts (hunk 6 above)
Comment (EzzhevNikita) [reply lib/Util.ts]: done

**Pull request merged**